### PR TITLE
go-controller: Sync services.

### PR DESCRIPTION
**Pull request by shettyg**
Signed-off-by: Gurucharan Shetty <guru@ovn.org>

### CODE REVIEW

File: go-controller/pkg/ovn/gateway.go [hunk 1]
```diff
@@ -17,6 +17,35 @@ func (ovn *Controller) getOvnGateways() ([]string, error) {
 	return strings.Fields(string(out)), err
 }
 
+func (ovn *Controller) getGatewayPhysicalIP(physicalGateway string) string {
```
Comment (dcbw) [inline go-controller/pkg/ovn/gateway.go]: Any reason to not just return an error from this function if the IP is ""?  Looks like most users have to C&P the error message code, might as well do it in one place.

File: go-controller/pkg/ovn/gateway.go [hunk 2]
```diff
@@ -17,6 +17,35 @@ func (ovn *Controller) getOvnGateways() ([]string, error) {
...
+		logrus.Errorf("Error in getting gateway physical IP from %s: %v(%v)",
+			physicalGateway, string(out), err)
+		return ""
+	}
+
+	physicalIP := strings.Trim(string(out), "\" \n")
+	return physicalIP
+}
+
+func (ovn *Controller) getGatewayLoadBalancer(physicalGateway,
```
Comment (dcbw) [inline go-controller/pkg/ovn/gateway.go]: Same here, I'd just return an error on "" loadbalancer IP.

File: go-controller/pkg/ovn/service.go [hunk 3]
```diff
@@ -1,10 +1,140 @@
...
+				continue
+			}
+
+			loadBalancerVIPS := ovn.getLoadBalancerVIPS(loadBalancer)
+			if loadBalancerVIPS == nil {
+				continue
+			}
+
+			for vip := range loadBalancerVIPS {
+				vipAndPort := strings.Split(vip, ":")
```
Comment (dcbw) [inline go-controller/pkg/ovn/service.go]: You can replace about 10 lines here with net.SplitHostPort() I think...

File: go-controller/pkg/ovn/gateway.go (hunk 1 above)
Comment (shettyg) [reply go-controller/pkg/ovn/gateway.go]: Makes sense. I decided to not print any errors in the function itself and leave that responsibility to caller.

File: go-controller/pkg/ovn/gateway.go (hunk 2 above)
Comment (shettyg) [reply go-controller/pkg/ovn/gateway.go]: same comment.

File: go-controller/pkg/ovn/service.go (hunk 3 above)
Comment (shettyg) [reply go-controller/pkg/ovn/service.go]: The SplitHostPort returns the port as a string too. So a conversion to int was needed. But, looks like we don't have to store ports as ints in the first place. So I used SplitHostPort and modified the datastructure to have ports stored as string.

**Comment by shettyg** [comment]
I gave it a round of testing. Please have a re-look.

File: go-controller/pkg/ovn/service.go [hunk 4]
```diff
@@ -1,10 +1,137 @@
...
+			logrus.Errorf("failed to get load-balancer vips for %s (%v)",
+				loadBalancer, err)
+			continue
+		}
+		if loadBalancerVIPS == nil {
+			continue
+		}
+
+		for vip := range loadBalancerVIPS {
+			if !stringSliceMembership(clusterServices[protocol], vip) {
```
Comment (rajatchopra) [inline go-controller/pkg/ovn/service.go]: An info/debug message about deleting the stale VIP would be useful.

Comment (shettyg) [reply go-controller/pkg/ovn/service.go]: I added a debug message.

**Pull request merged**